src/com/xilinx/rapidwright/router/RouteThruHelper.java

```diff
@@ -158,21 +158,57 @@ public static boolean isRouteThruPIPAvailable(Design design, PIP routethru) {
         return isRouteThruPIPAvailable(design, routethru.getStartWire(), routethru.getEndWire());
     }
 
+    private static boolean isRouteThruSitePinAvailable(Design design, SitePin sitePin) {
+        if (sitePin == null) {
+            return false;
+        }
+        SiteInst siteInst = design.getSiteInstFromSite(sitePin.getSite());
+        if (siteInst == null) {
+            return true;
+        }
+        Net netCollision = siteInst.getNetFromSiteWire(sitePin.getBELPin().getSiteWireName());
+        if (netCollision != null) {
+            return false;
+        }
+        return true;
+    }
+
     public static boolean isRouteThruPIPAvailable(Design design, Wire start, Wire end) {
         SitePin outPin = end.getSitePin();
-        if (outPin == null) return false;
-        SiteInst siteInst = design.getSiteInstFromSite(outPin.getSite());
-        if (siteInst == null) return true;
-        Net outputNetCollision = siteInst.getNetFromSiteWire(outPin.getBELPin().getSiteWireName());
-        if (outputNetCollision != null) return false;
+        if (!isRouteThruSitePinAvailable(design, outPin)) {
+            return false;
+        }
         SitePin inPin = start.getSitePin();
-        BELPin belPin = inPin.getBELPin();
-        Net inputNetCollision = siteInst.getNetFromSiteWire(belPin.getSiteWireName());
-        if (inputNetCollision != null) return false;
+        assert(inPin.getSite() == outPin.getSite());
+        if (!isRouteThruSitePinAvailable(design, inPin)) {
+            return false;
+        }
+        SiteInst siteInst = design.getSiteInstFromSite(inPin.getSite());
+        for (BELPin sink : inPin.getBELPin().getSiteConns()) {
+            Cell cellCollision = siteInst.getCell(sink.getBEL());
+            if (cellCollision != null) {
+                return false;
+            }
+        }
+        return true;
+    }
 
-        for (BELPin sink : belPin.getSiteConns()) {
-            Cell collision = siteInst.getCell(sink.getBEL());
-            if (collision != null) return false;
+    public static boolean isRouteThruPIPAvailable(Design design, Node start, Node end) {
```

Comment:
Would it make sense to unify this method with the one above, have a backend method like `public static boolean isRouteThruPIPAvailable(Design design, SitePin start, SitePin end)`?  The method bodies are very similar.

Reply:
The reason I avoided doing this is because computing the second SitePin is expensive, and we'd like to avoid doing that if possible. The other option would be to take a `Tile`+`int` instead of a `Wire` or `Node` since converting either of the latter to the other would require allocation.

Reply:
Sure, that make sense.  Let's leave it alone for now.  The implementation of `Node.getSitePin()` would likely require another allocation to make that work.  Maybe just drop a comment that the method bodies should be kept in sync since they are doing the same thing.  Perhaps we could introduce an interface that would be compatible for both `Node` and `Wire` objects to avoid this in the future.